EIPS/eip-00.md

```diff
@@ -0,0 +1,831 @@
+
+```solidity
+
+function validateTransaction(uint256 version, bytes32 txHash, bytes transaction) external returns (uint256 validationData);
+
+```
+
+The gas limit of this frame is set to `validationGasLimit - senderCreationGasUsed`.\
+The `transaction` parameter is interpreted as an ABI encoding of `TransactionType4`.\
+The `txHash` parameter represents the hash of the AA_TX_TYPE transaction with empty signature, as defined in section
+[Calculation of Transaction Type AA_TX_TYPE hash](#calculation-of-transaction-type-aatxtype-hash).\
+The `version` parameter is added in order to maintain the Solidity method ID in case of changes to this struct
+in future revisions of this EIP.
+
+The amount of gas used by this frame is referred to as `senderValidationGasUsed`.
+
+The frame must return 32 bytes `validationData` that is interpreted as:
+
+```solidity
+
+abi.encodePacked(MAGIC_VALUE_SENDER, validUntil, validAfter)
+
+```
+
+In order to allow a gas estimation to determine the amount of gas that this frame
+requires to complete successfully while not having the actual `signature` value, this function
+should avoid reverting on invalid signature, and should return a value different from `MAGIC_VALUE_SENDER`.
+
+Type of the `validUntil` is 6-byte timestamp value, or zero for "infinite". The transaction is valid only up to this time.
+Type of the `validAfter` is 6-byte timestamp. The transaction is valid only after this time.
+
+The `validateTransaction` function can choose to revert on any condition that can be satisfied during gas estimation.
+
+#### Paymaster validation frame
+
+The `paymaster` of the transaction, if specified, is invoked with the following data:
+
+```solidity
+
+function validatePaymasterTransaction(uint256 version, bytes32 txHash, bytes transaction) external returns (bytes context, uint256 validationData);
+
+```
+
+The gas limit of this frame is set to `paymasterGasLimit`.
+
+The amount of gas used by this frame is referred to as `paymasterValidationGasUsed`.
+
+The `transaction` parameter is interpreted as an ABI encoding of `TransactionType4`.\
+The `txHash` parameter represents the hash of the AA_TX_TYPE transaction with empty signature, as defined in section
+[Calculation of Transaction Type AA_TX_TYPE hash](#calculation-of-transaction-type-aatxtype-hash).
+
+The frame must return a bytes array that is interpreted as:
+
+```solidity
+
+abi.encode(context, MAGIC_VALUE_PAYMASTER, validUntil, validAfter)
+
+```
+
+Same as in the [`sender` validation frame](#sender-validation-frame), in order to support gas estimation this
+frame should return a value different from `MAGIC_VALUE_PAYMASTER` for conditions that cannot be satisfied
+before signing.
+
+The size of the `context` byte array may not exceed `MAX_CONTEXT_SIZE` for a transaction to be considered valid.
+
+#### Sender execution frame
+
+The `sender` address is invoked with `callData` input.
+
+The gas limit of this frame is set to `callGasLimit - calldataCost`.\
```

Comment:
why not cover calldataCost during validation? it would simplify execution
(e.g.: when calculating the callDataCost, it is strange that it depends on whether this is the 1st transaction (which also delpoys) or not.

Reply:
If we give the entire `calldataCost` to this frame, what happens if it actually uses more than `callGasLimit - calldataCost`, e.g. use exactly `calldataCost`?  It means that the total cost of this validation is now `callGasLimit + calldataCost`.